compiler/rustc_typeck/src/check/expr.rs

```diff
@@ -1030,24 +1061,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let count = self.to_const(count);
 
         let uty = match expected {
-            ExpectHasType(uty) => match *uty.kind() {
+            ExpectHasType(uty, _) => match *uty.kind() {
                 ty::Array(ty, _) | ty::Slice(ty) => Some(ty),
                 _ => None,
             },
             _ => None,
         };
 
+        let type_ascr_ctxt = expected.get_coercion_ctxt();
         let (element_ty, t) = match uty {
             Some(uty) => {
-                self.check_expr_coercable_to_type(&element, uty, None);
+                self.check_expr_coercable_to_type(&element, uty, None, type_ascr_ctxt);
                 (uty, uty)
             }
             None => {
                 let ty = self.next_ty_var(TypeVariableOrigin {
                     kind: TypeVariableOriginKind::MiscVariable,
                     span: element.span,
                 });
-                let element_ty = self.check_expr_has_type_or_error(&element, ty, |_| {});
+                let element_ty =
+                    self.check_expr_has_type_or_error(&element, ty, |_| {}, type_ascr_ctxt);
```

Comment:
Hmm, this is a sort of tricky case. This corresponds to `[E; 22]` and I guess that we want the expression `E` to always be considered a coercion site. So this should probably just be a call to `check_Expr_coerces_to` but using the target type of a fresh variable. But I'm not sure why the code is written this way, there may be a reason, I guess it has to do with extracting as much info about the element type as we can. Still, I would prefer to just simplify to always call `coercable_To`